File: apps/agent/src/local-proxy.ts
import { request as httpRequest } from "node:http";

import WebSocket from "ws";

export interface ProxyRequestMessage {
  type: "proxy_request";
  requestId: string;
  localPort: number;
  method: string;
  path: string;
  headers: Record<string, string | string[]>;
  body: string;
}

export interface WebSocketConnectMessage {
  type: "websocket_connect";
  connectionId: string;
  localPort: number;
  path: string;
  headers: Record<string, string | string[]>;
  protocols: string[];
}

export interface WebSocketConnectedMessage {
  type: "websocket_connected";
  connectionId: string;
  protocol?: string;
}

export interface WebSocketConnectErrorMessage {
  type: "websocket_connect_error";
  connectionId: string;
  message: string;
}

export interface WebSocketDataMessage {
  type: "websocket_data";
  connectionId: string;
  data: string;
  isBinary: boolean;
}

export interface WebSocketCloseMessage {
  type: "websocket_close";
  connectionId: string;
  code?: number;
  reason?: string;
}

const HOP_BY_HOP_HEADERS = new Set([
  "accept-encoding",
  "connection",
  "content-length",
  "host",
  "keep-alive",
  "proxy-connection",
  "te",
  "trailer",
  "transfer-encoding",
  "upgrade",
]);

const STRIPPED_WEBSOCKET_HEADERS = new Set([
  "sec-websocket-extensions",
  "sec-websocket-key",
  "sec-websocket-protocol",
  "sec-websocket-version",
]);

function readHeaderValue(value: string | string[]): string {
  return Array.isArray(value) ? value.join(", ") : value;
}

export function buildLocalProxyHeaders(
  headers: Record<string, string | string[]>,
): Record<string, string> {
  const result: Record<string, string> = {};
  let originalHost: string | undefined;

  for (const [key, value] of Object.entries(headers)) {
    const normalizedKey = key.toLowerCase();

    if (normalizedKey === "x-bore-original-host") {
      const candidate = readHeaderValue(value).trim();

      if (candidate) {
        originalHost = candidate;
      }

      continue;
    }

    if (HOP_BY_HOP_HEADERS.has(normalizedKey)) {
      continue;
    }

    result[key] = readHeaderValue(value);
  }

  if (originalHost) {
    result.Host = originalHost;
  }

  return result;
}

export function buildLocalWebSocketHeaders(
  headers: Record<string, string | string[]>,
): Record<string, string> {
  const result = buildLocalProxyHeaders(headers);

  for (const key of Object.keys(result)) {
    if (STRIPPED_WEBSOCKET_HEADERS.has(key.toLowerCase())) {
      delete result[key];
    }
  }

  return result;
}

export async function connectLocalWebSocket(
  message: WebSocketConnectMessage,
): Promise<{ socket: WebSocket; protocol?: string }> {
  const target = new URL(message.path, `ws://127.0.0.1:${message.localPort}`);
  const requestedProtocols =
    message.protocols.length === 0
      ? undefined
      : message.protocols.length === 1
        ? message.protocols[0]
        : message.protocols;

  return new Promise((resolve, reject) => {
    const socket = new WebSocket(target, requestedProtocols, {
      headers: buildLocalWebSocketHeaders(message.headers),
    });
    let settled = false;

    const settle = (callback: () => void) => {
      if (settled) {
        return;
      }

      settled = true;
      socket.off("open", handleOpen);
      socket.off("error", handleError);
      socket.off("close", handleClose);
      socket.off("unexpected-response", handleUnexpectedResponse);
      callback();
    };

    const handleOpen = () => {
      settle(() => {
        resolve({
          socket,
          protocol: socket.protocol || undefined,
        });
      });
    };

    const handleError = (error: Error) => {
      settle(() => reject(error));
    };

    const handleClose = () => {
      settle(() => reject(new Error("Local websocket closed during connect")));
    };

    const handleUnexpectedResponse = (_request: unknown, response: import("node:http").IncomingMessage) => {
      const chunks: Buffer[] = [];

      response.on("data", (chunk) => {
        chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
      });
      response.on("end", () => {
        const detail = Buffer.concat(chunks).toString("utf8").trim();
        const statusCode = response.statusCode ?? 502;
        const suffix = detail ? `: ${detail}` : "";
        settle(() => reject(new Error(`Local websocket upgrade failed with status ${statusCode}${suffix}`)));
      });
    };

    socket.on("open", handleOpen);
    socket.on("error", handleError);
    socket.on("close", handleClose);
    socket.on("unexpected-response", handleUnexpectedResponse);
  });
}

export async function proxyLocalRequest(message: ProxyRequestMessage) {
  const body =
    message.method === "GET" || message.method === "HEAD"
      ? undefined
      : Buffer.from(message.body, "base64");

  const response = await new Promise<{
    status: number;
    headers: Record<string, string>;
    body: string;
  }>((resolve, reject) => {
    const request = httpRequest(
      {
        host: "127.0.0.1",
        port: message.localPort,
        method: message.method,
        path: message.path,
        headers: buildLocalProxyHeaders(message.headers),
      },
      (response) => {
        const chunks: Buffer[] = [];

        response.on("data", (chunk) => {
          chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
        });
        response.on("end", () => {
          const headers: Record<string, string> = {};

          for (const [key, value] of Object.entries(response.headers)) {
            if (!value || ["content-encoding", "content-length"].includes(key.toLowerCase())) {
              continue;
            }

            headers[key] = Array.isArray(value) ? value.join(", ") : value;
          }

          resolve({
            status: response.statusCode ?? 502,
            headers,
            body: Buffer.concat(chunks).toString("base64"),
          });
        });
      },
    );

    request.on("error", reject);

    if (body) {
      request.write(body);
    }

    request.end();
  });

  return {
    type: "proxy_response" as const,
    requestId: message.requestId,
    status: response.status,
    headers: response.headers,
    body: response.body,
  };
}

File: apps/agent/internal/agent/localproxy.go
package agent

import (
	"bytes"
	"context"
	"encoding/base64"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/gorilla/websocket"
)

var hopByHopHeaders = map[string]struct{}{
	"accept-encoding":   {},
	"connection":        {},
	"content-length":    {},
	"host":              {},
	"keep-alive":        {},
	"proxy-connection":  {},
	"te":                {},
	"trailer":           {},
	"transfer-encoding": {},
	"upgrade":           {},
}

var strippedWebSocketHeaders = map[string]struct{}{
	"sec-websocket-extensions": {},
	"sec-websocket-key":        {},
	"sec-websocket-protocol":   {},
	"sec-websocket-version":    {},
}

type proxyHeaders struct {
	headers http.Header
	host    string
}

type localWebSocket struct {
	conn      *websocket.Conn
	writeLock chan struct{}
}

func newLocalWebSocket(conn *websocket.Conn) *localWebSocket {
	lock := make(chan struct{}, 1)
	lock <- struct{}{}
	return &localWebSocket{conn: conn, writeLock: lock}
}

func (socket *localWebSocket) write(messageType int, payload []byte) error {
	<-socket.writeLock
	defer func() {
		socket.writeLock <- struct{}{}
	}()

	return socket.conn.WriteMessage(messageType, payload)
}

func (socket *localWebSocket) close(code int, reason string) error {
	<-socket.writeLock
	defer func() {
		socket.writeLock <- struct{}{}
	}()

	if code == 0 {
		return socket.conn.Close()
	}

	message := websocket.FormatCloseMessage(code, trimCloseReason(reason))
	return socket.conn.WriteControl(websocket.CloseMessage, message, time.Now().Add(time.Second))
}

func trimCloseReason(reason string) string {
	if len(reason) > 123 {
		return reason[:123]
	}
	return reason
}

func buildLocalProxyHeaders(headers map[string][]string) proxyHeaders {
	result := make(http.Header)
	var originalHost string

	for key, values := range headers {
		normalized := strings.ToLower(key)

		if normalized == "x-bore-original-host" {
			if len(values) > 0 {
				originalHost = strings.TrimSpace(values[0])
			}
			continue
		}

		if _, blocked := hopByHopHeaders[normalized]; blocked {
			continue
		}

		for _, value := range values {
			result.Add(key, value)
		}
	}

	return proxyHeaders{
		headers: result,
		host:    originalHost,
	}
}

func buildLocalWebSocketHeaders(headers map[string][]string) proxyHeaders {
	result := buildLocalProxyHeaders(headers)

	for key := range result.headers {
		if _, blocked := strippedWebSocketHeaders[strings.ToLower(key)]; blocked {
			result.headers.Del(key)
		}
	}

	return result
}

func proxyLocalRequest(message proxyRequestMessage) (proxyResponseMessage, error) {
	target := &url.URL{
		Scheme: "http",
		Host:   fmt.Sprintf("127.0.0.1:%d", message.LocalPort),
		Path:   "/",
	}

	parsedPath, err := url.Parse(message.Path)
	if err != nil {
		return proxyResponseMessage{}, err
	}

	target.Path = parsedPath.Path
	target.RawPath = parsedPath.RawPath
	target.RawQuery = parsedPath.RawQuery

	var body io.Reader
	if message.Method != http.MethodGet && message.Method != http.MethodHead && message.Body != "" {
		decoded, err := base64.StdEncoding.DecodeString(message.Body)
		if err != nil {
			return proxyResponseMessage{}, err
		}
		body = bytes.NewReader(decoded)
	}

	req, err := http.NewRequest(message.Method, target.String(), body)
	if err != nil {
		return proxyResponseMessage{}, err
	}

	headers := buildLocalProxyHeaders(message.Headers)
	req.Header = headers.headers
	if headers.host != "" {
		req.Host = headers.host
	}

	res, err := (&http.Client{Timeout: 30 * time.Second}).Do(req)
	if err != nil {
		return proxyResponseMessage{}, err
	}
	defer res.Body.Close()

	responseBody, err := io.ReadAll(res.Body)
	if err != nil {
		return proxyResponseMessage{}, err
	}

	responseHeaders := make(map[string][]string)
	for key, values := range res.Header {
		switch strings.ToLower(key) {
		case "content-encoding", "content-length":
			continue
		}

		responseHeaders[key] = append([]string(nil), values...)
	}

	return proxyResponseMessage{
		Type:      "proxy_response",
		RequestID: message.RequestID,
		Status:    res.StatusCode,
		Headers:   responseHeaders,
		Body:      base64.StdEncoding.EncodeToString(responseBody),
	}, nil
}

func connectLocalWebSocket(ctx context.Context, message websocketConnectMessage) (*localWebSocket, string, error) {
	target := &url.URL{
		Scheme: "ws",
		Host:   fmt.Sprintf("127.0.0.1:%d", message.LocalPort),
		Path:   "/",
	}

	parsedPath, err := url.Parse(message.Path)
	if err != nil {
		return nil, "", err
	}

	target.Path = parsedPath.Path
	target.RawPath = parsedPath.RawPath
	target.RawQuery = parsedPath.RawQuery

	headers := buildLocalWebSocketHeaders(message.Headers)
	if headers.host != "" {
		headers.headers.Set("Host", headers.host)
	}

	dialer := websocket.Dialer{
		HandshakeTimeout: 10 * time.Second,
		Subprotocols:     message.Protocols,
	}

	conn, response, err := dialer.DialContext(ctx, target.String(), headers.headers)
	if err != nil {
		if response != nil && response.Body != nil {
			detail, _ := io.ReadAll(response.Body)
			_ = response.Body.Close()
			suffix := strings.TrimSpace(string(detail))
			if suffix != "" {
				return nil, "", fmt.Errorf("local websocket upgrade failed with status %d: %s", response.StatusCode, suffix)
			}
			return nil, "", fmt.Errorf("local websocket upgrade failed with status %d", response.StatusCode)
		}

		return nil, "", err
	}

	return newLocalWebSocket(conn), conn.Subprotocol(), nil
}
